src/Entity/MetadataDisplayEntity.php

```diff
@@ -438,30 +443,71 @@ public function getTwigVariablesUsed() {
-        $value = $node->getAttribute('value');
-        if (!empty($value) && is_string($value)) {
-          $variables[$value] = $value;
+      $lineno = [$node->getTemplateLine()];
+      $variable_key = '';
+      $parent_path = '';
+      if ($node instanceof NameExpression && !$nodes instanceof NotUnary) {
+        $variable_key = $node->getAttribute('name');
+        if($variable_key == '_key') {
+          $seq = $nodes->getNode('seq');
+          $seq_name = $seq->hasNode('node') ? $seq->getNode('node')->getAttribute('name') : $seq->getAttribute('name');
+          $seq_value = $seq->hasNode('attribute') ? $seq->getNode('attribute')->getAttribute('value') : '';
+          $variable_key = $nodes->getNode('value_target')->getAttribute('name');
+          $parent_path = empty($seq_value) ? $seq_name : $seq_name . '.' . $seq_value;
+          $set_var = $variable_key;
+          $set_source = $parent_path;
         }
       }
-      elseif ($node instanceof \Twig\Node\Expression\GetAttrExpression) {
-        $path = implode('.', $this->getTwigVariableNames($node));
-        if (!empty($path)) {
-          $variables[$path] = $path;
+      elseif ($node instanceof ConstantExpression && $nodes instanceof GetAttrExpression) {
+        $variable_key = $node->getAttribute('value');
+      }
+      elseif ($node instanceof GetAttrExpression) {
+        $variable_names = $this->getTwigVariableNames($node, array_replace_recursive($all_variables, $variables), $set_var, $set_source);
+        $variable_names_flat = [];
+        $prev_name = null;
+        foreach ($variable_names as $name) {
+          $name_check = $prev_name ? $prev_name['path'] . '.' . $name['path']:'';
+          if (($name['path'] == $set_var) && !($set_source == $name_check)) {
+            array_push($variable_names_flat, $set_source);
+          }
+          else {
+            array_push($variable_names_flat, $name['path']);
+          }
+          $prev_name = $name;
         }
+        $variable_key = implode('.', $variable_names_flat);
       }
-      elseif ($node instanceof \Twig\Node\Node) {
-        $variables += $this->getTwigVariableNames($node);
+      elseif ($node instanceof Node) {
+        $add_variables = $this->getTwigVariableNames($node, array_replace_recursive($all_variables, $variables), $set_var, $set_source);
+        $variables = array_replace_recursive($variables, $add_variables);
+      }
+      if (!empty($variable_key) && is_string($variable_key)) {
+        $variables[$variable_key]['path'] = $variable_key;
+        if(isset($all_variables[$variable_key]['line'])) {
+          $variables[$variable_key]['line']     = array_unique(array_merge($all_variables[$variable_key]['line'], $lineno));
+          $all_variables[$variable_key]['line'] = array_unique(array_merge($all_variables[$variable_key]['line'], $lineno));
+        }
+        if(!isset($variables[$variable_key]['line'])) {
+          $variables[$variable_key]['line'] = $lineno;
+        }
```

Comment:
This piece here looks strange. So if only present in a single line you make ['line'] a string, if not it is an array. Why not make it an array always? That is more consistent when parsing it on the caller. Also 
```
$variables[$variable_key]['line']     = array_unique(array_merge($all_variables[$variable_key]['line'], $lineno));
           $all_variables[$variable_key]['line'] = array_unique(array_merge($all_variables[$variable_key]['line'], $lineno));
           ````
Is doing the same exactly twice.  I don't see you use $all_variables at all afterwards not return it, but if you need it 
$variables[$variable_key]['line'] = $all_variables[$variable_key]['line'] = array_unique(array_merge($all_variables[$variable_key]['line'], $lineno));

Reply:
>This piece here looks strange. So if only present in a single line you make ['line'] a string, if not it is an array. Why not make it an array always? That is more consistent when parsing it on the caller.

`$lineno = [$node->getTemplateLine()];` above it makes it always an array, no? Did I miss something?

Reply:
Re: the `$all_variables`, will update to your suggestion for now (didn't know you could chain assignments, cool!) and address the larger recursion issue in the below comment.